src/app/loadout-builder/process/useProcess.ts

```diff
@@ -230,158 +219,199 @@ function createWorker() {
-    const statValues: number[] = [];
-    const statsByHash = item.stats && keyByStatHash(item.stats);
-    // Ensure ordering of stats
-    // TODO: statOrder includes disabled stats, should we omit them?
-    if (statsByHash) {
-      const assumedEnergy = calculateAssumedItemEnergy(item, armorEnergyRules);
-      for (const statHash of statOrder) {
-        let value = statsByHash[statHash]!.base;
-        // Add in masterwork stat bonus if we're assuming masterwork stats
-        if (assumedEnergy === MAX_ARMOR_ENERGY_CAPACITY) {
-          value += 2;
-        }
-        statValues.push(value);
-      }
-      // Also use assumed energy because a class item with 9 energy is better than one with 8
-      statValues.push(assumedEnergy);
-    }
-    statsCache.set(item, statValues);
+  // Second pass -- cache the worker-relevant information, except the one we used in the first pass.
+  const cache = new Map<
+    DimItem,
+    { stats: number[]; energyCapacity: number; relevantModSeasons: Set<string> }
+  >();
+  for (const item of mappedItems) {
+    // Id, name are not important, exoticness+hash and energy type were grouped by in phase 1.
+    // Energy value is the same for all items.
+
+    // Item stats are important for the stat results of a full set
+    const statValues: number[] = statOrder.map((s) => item.processItem.stats[s]);
+    // Energy capacity affects mod assignment
+    const energyCapacity = item.processItem.energy?.capacity || 0;
+    // Supported mod tags affect mod assignment
+    const relevantModSeasons =
+      item.processItem.compatibleModSeasons?.filter((season) => requiredModTags.has(season)) ?? [];
+    relevantModSeasons.sort();
+
+    cache.set(item.dimItem, {
+      stats: statValues,
+      energyCapacity,
+      relevantModSeasons: new Set(relevantModSeasons),
+    });
   }
 
-  // Group items by their exact stats. The statsCache includes mod energy capacity
-  // but that's fine because items in the same group with the same stats but lower
-  // energy capacities already got excluded.
-  const statGroupingFn = (item: DimItem) => statsCache.get(item)!.toString();
+  // Group items by everything relevant.
+  const finalGroupingFn = (item: DimItem) => {
+    const info = cache.get(item)!;
+    return `${info.stats}-${info.energyCapacity}-${[...info.relevantModSeasons.values()]}`;
+  };
 
-  const modGroups = _.groupBy(items, modGroupingFn);
+  const energyGroups = _.groupBy(mappedItems, ({ processItem }) =>
+    firstPassGroupingFn(processItem)
+  );
 
-  // Final grouping by both mod requirements AND exact stats
-  const groups: DimItem[][] = [];
+  // Final grouping by everything relevant
+  const groups: ItemGroup[] = [];
 
-  // Go through each grouping-by-mod, throw out any items with worse stats than
-  // another item in that group, then use what's left to build groups by exact stats.
-  for (const group of Object.values(modGroups)) {
-    const statSet = new StatsSet<DimItem>();
```

Comment:
The idea of the StatsSet was to be more efficient than the n^2 comparison of all items against all other items - do you think we don't need that optimization?

Reply:
I didn't notice any slow down as a result but will do more profiling.

Reply:
So I ran this with my 15-20 legendary items per bucket and this is what it looked like before:

![grafik](https://user-images.githubusercontent.com/14299449/182391686-a9d6d4ed-2a97-4528-bb26-b8ab1bc314f8.png)

and after:

![grafik](https://user-images.githubusercontent.com/14299449/182388854-def2371e-f052-4df5-84c4-fa79efa35144.png)

Seems like a wash. The new code also does more things to maintain correctness, so I'm not seeing a performance concern here.